src/nvim/message.c

```diff
@@ -3064,14 +3077,24 @@ int msg_end(void)
 
 void msg_ext_ui_flush(void)
 {
-  if (!ui_has(kUIMessages)) {
+  if (!msg_enable_ext()) {
     return;
   }
 
   msg_ext_emit_chunk();
   if (msg_ext_chunks.size > 0) {
-    ui_call_msg_show(cstr_to_string(msg_ext_kind),
-                     msg_ext_chunks, msg_ext_overwrite);
+    if (ui_has(kUIMessages)) {
+      ui_call_msg_show(cstr_to_string(msg_ext_kind),
+                       msg_ext_chunks, msg_ext_overwrite);
+    } else {
+      msg_call_msgfunc("msg_show", &msg_ext_chunks);
+
+      // ui_call_msg_xxx() calls ui_event(), which frees the allocated chunk
+      // The allocated chunk should also need freeing in or after
+      // msg_call_msgfunc().
+      api_free_array(msg_ext_chunks);
```

Comment:
Can this be deferred to `msg_call_msgfunc_event`?

Reply:
OK.  It should be.

Reply:
Do not deferred the release of `msg_ext_chunks`. If `msg_ext_emit_chunk` is called before `msg_call_msgfunc_event` is called, a memory leak will occur. Since the string is newly allocated in `object_to_vim` ([lines 3212-3213](https://github.com/neovim/neovim/pull/16480/files#diff-0fdec949ebc770643273989f95da16a0d9bee6b68db894fe9ea61a8025397a15R3212-R3213)), there is no problem freeing it here.

Reply:
Makes sense. I was not sure if the data was being copied before being usage on `msg_call_msgfunc_event`.

It's good to explain the motivation in the comment

```c
// since msg_ext_chunks has been copied in msg_call_msgfunc, it can already be
// freed here
       api_free_array(msg_ext_chunks); 
```

Reply:
OK.